iroh-resolver/src/balanced_tree.rs

```diff
@@ -0,0 +1,451 @@
+                // at this point the tree will be able to recieve new links
+                // without "overflowing", aka the leaf node and stem nodes
+                // have fewer than `degree` number of links
+            }
+
+            // now that we know the tree is in a "healthy" state to
+            // recieve more links, add the link to the tree
+            let (cid, bytes, len) = TreeNode::Leaf(chunk).encode()?;
+            tree[0].push((cid, len));
+            yield (cid, bytes);
+            // at this point, the leaf node may have `degree` number of
+            // links, but no other stem node will
+        }
+
+        // our stream had 1 chunk that we have already yielded
+        if tree.len() == 1 && tree[0].len() == 1 {
+            return
+        }
+
+        // clean up, aka yield the rest of the stem nodes
+        // since all the stem nodes are able to recieve links
+        // we don't have to worry about "overflow"
+        while let Some(links) = tree.pop_front() {
+            let (cid, bytes, len) = TreeNode::Stem(links).encode()?;
+            yield (cid, bytes);
+
+            if let Some(front) = tree.front_mut() {
+                front.push((cid, len));
+            } else {
+                // final root, nothing to do
+            }
+        }
+    }
+}
+
+fn create_unixfs_node_from_links(links: Vec<(Cid, u64)>, is_root: bool) -> Result<UnixfsNode> {
+    let links = links
+        .into_iter()
+        .map(|(cid, len)| dag_pb::PbLink {
+            hash: Some(cid.to_bytes()),
+            name: None,
+            tsize: Some(len as u64),
+        })
+        .collect();
+
+    // PBNode.Data
+    let inner = unixfs_pb::Data {
+        r#type: DataType::File as i32,
+        ..Default::default()
+    };
+
+    // create PBNode
+    let outer = encode_unixfs_pb(&inner, links)?;
+
+    if is_root {
+        return Ok(UnixfsNode::Directory(Node { inner, outer }));
+    }
+
+    // create UnixfsNode
+    Ok(UnixfsNode::File(Node { inner, outer }))
+}
+
+// Leaf and Stem nodes are the two types of nodes that can exist in the tree
+// Leaf nodes encode to `UnixfsNode::Raw`
+// Stem nodes encode to `UnixfsNode::File`
+enum TreeNode {
+    Leaf(Bytes),
+    Stem(Vec<(Cid, u64)>),
+}
+
+impl TreeNode {
+    fn encode(self) -> Result<(Cid, Bytes, u64)> {
+        match self {
+            TreeNode::Leaf(bytes) => {
+                let node = UnixfsNode::Raw(bytes);
+                let (cid, bytes) = node.encode()?;
+                let len = bytes.len();
+                Ok((cid, bytes, len as u64))
+            }
+            TreeNode::Stem(links) => {
+                // keep track of `tsize`, aka the size of the encoded tree at the given link
+                let mut cumulative_len: u64 = links.iter().map(|(_, len)| len).sum();
+                let node = create_unixfs_node_from_links(links, false)?;
+                let (cid, bytes) = node.encode()?;
+                cumulative_len += bytes.len() as u64;
```

Comment:
I am not sure if this length calculation is correct, @ribasushi thoughts?

Reply:
yes please a million more eyes on this. 

https://github.com/ipfs/kubo/blob/b3faaad1310bcc32dc3dd24e1919e9edf51edba8/merkledag/pb/merkledag.proto#L27-L28

took this to mean: cumulative size of the tree that this link points to. Which would include the size of the encoded node that link points to + the tsize of all of that node's links.

Reply:
@ramfox @dignifiedquire this value is effectively "freeform". You can omit it, or put whatever you feel like in there ( full DAG weight, deduplicated DAG weight, the payload-sum-size, whatever ). Aside from rendering "directory sizes" on gateways this value has no other function.

Reply:
okay @dignifiedquire what's iroh's opinion on what this value should be?

Reply:
I think yours is fine for now, we might change our opinion later